include/fiddle/mechanics/force_contribution_lib.h

```diff
@@ -564,6 +565,194 @@ namespace fdl
+                                      const Tensor<1, spacedim, Number> fiber_i)
+  {
+    std::vector<double> I4_i(CC.size());
+    for(unsigned int i = 0; i < CC.size(); ++i)
+      I4_i[i] = (CC[i] * fiber_i) * fiber_i;
+    return I4_i;
+  }
+
+  /*
+   * dI4_i_dFF for one quadrature point
+   */
+  template<int spacedim, typename Number = double>
+  inline Tensor<2, spacedim, Number> dI4_i_dFF(
+    const Tensor<2, spacedim, Number> FF,
+    const Tensor<1, spacedim, Number> fiber_i)
+  {
+    return 2.0 * outer_product(FF * fiber_i, fiber_i);
+  }
+
+  /*
+   * dI4_i_dFF for all quadrature points
+   */
+  template<int spacedim, typename Number = double>
+  inline std::vector<Tensor<2, spacedim, Number>> dI4_i_dFF_vec(
+    const std::vector<Tensor<2, spacedim, Number>> FF,
+    const Tensor<1, spacedim, Number> fiber_i)
+  {
+    std::vector<Tensor<2, spacedim, Number>> dI4_i_dFF(FF.size());
+    for(unsigned int i = 0; i < FF.size(); ++i)
+      dI4_i_dFF[i] = 2.0 * outer_product(FF[i] * fiber_i, fiber_i);
+    return dI4_i_dFF;
+  }
+
+  /*
+   * I8_ij for one quadrature point
+   */
+  template<int spacedim, typename Number = double>
+  inline double I8_ij(const SymmetricTensor<2, spacedim, Number> CC,
+                      const Tensor<1, spacedim, Number>          fiber_i,
+                      const Tensor<1, spacedim, Number>          fiber_j)
+  {
+    return (CC * fiber_j) * fiber_i;
+  }
+
+  /*
+   * I8_ij for all quadrature points
+   */
+  template<int spacedim, typename Number = double>
+  inline std::vector<double> I8_ij_vec(const std::vector<SymmetricTensor<2, spacedim, Number>> CC,
+                                       const Tensor<1, spacedim, Number> fiber_i,
+                                       const Tensor<1, spacedim, Number> fiber_j)
+  {
+    std::vector<double> I8_ij(CC.size());
+    for(unsigned int i = 0; i < CC.size(); ++i)
+      I8_ij[i] = (CC[i] * fiber_j) * fiber_i;
+    return I8_ij;
+  }
+
+  /*
+   * dI8_ij_dFF for one quadrature point
+   */
+  template<int spacedim, typename Number = double>
+  inline Tensor<2, spacedim, Number> dI8_ij_dFF(
+    const Tensor<2, spacedim, Number> FF,
+    const Tensor<1, spacedim, Number> fiber_i,
+    const Tensor<1, spacedim, Number> fiber_j)
+  {
+    return outer_product(FF * fiber_j, fiber_i) + outer_product(FF * fiber_i, fiber_j);
+  }
+
+  /*
+   * dI8_ij_dFF for all quadrature points
+   */
+  template<int spacedim, typename Number = double>
+  inline std::vector<Tensor<2, spacedim, Number>> dI8_ij_dFF_vec(
+    const std::vector<Tensor<2, spacedim, Number>> FF,
+    const Tensor<1, spacedim, Number> fiber_i,
+    const Tensor<1, spacedim, Number> fiber_j)
+  {
+    std::vector<Tensor<2, spacedim, Number>> dI8_ij_dFF(FF.size());
+    for(unsigned int i = 0; i < FF.size(); ++i)
+      dI8_ij_dFF[i] = outer_product(FF[i] * fiber_j, fiber_i) + outer_product(FF[i] * fiber_i, fiber_j);
+    return dI8_ij_dFF;
+  }
```

Comment:
if these functions allocate memory they are less likely to be inlined - we should avoid doing it this way (also allocating memory may be relatively expensive).

An alternative would be to define `FF`, `f`, and `s` with aliases so that things are not so verbose.